libdrgn/program.c

```diff
@@ -1108,6 +1108,140 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
 	return NULL;
 }
 
+DEFINE_VECTOR(symbolp_vector, struct drgn_symbol *);
+
+enum {
+	SEARCH_NAME = (1 << 0),
+	SEARCH_ADDRESS = (1 << 1),
+	SEARCH_ALL = (1 << 2),
+};
+
+struct symbols_search_arg {
+	const char *name;
+	uint64_t address;
+	struct symbolp_vector results;
+	struct drgn_error *err;
+	bool bad_symtabs;
+	unsigned int flags;
+};
+
+static bool symbol_match(struct symbols_search_arg *arg, GElf_Addr addr,
+			 const GElf_Sym *sym, const char *name)
+{
+	if (arg->flags & SEARCH_ALL)
+		return true;
+	if (arg->flags & SEARCH_NAME)
+		if (name && arg->name && strcmp(name, arg->name) == 0)
+			return true;
+	if (arg->flags & SEARCH_ADDRESS)
+		if (arg->address >= addr && arg->address < addr + sym->st_size)
+			return true;
+	return false;
+}
+
+static int symbols_search_cb(Dwfl_Module *dwfl_module, void **userdatap,
+			     const char *module_name, Dwarf_Addr base,
+			     void *cb_arg)
+{
+	struct symbols_search_arg *arg = cb_arg;
+	int symtab_len, i;
+
+	symtab_len = dwfl_module_getsymtab(dwfl_module);
+	if (symtab_len == -1) {
+		arg->bad_symtabs = true;
+		return DWARF_CB_OK;
+	}
+	/* Ignore the zeroth null symbol */
+	for (i = 1; i < symtab_len; i++) {
+		GElf_Sym elf_sym;
+		GElf_Addr elf_addr;
+		const char *name;
+		struct drgn_symbol *sym;
+
+		name = dwfl_module_getsym_info(dwfl_module, i, &elf_sym,
+					       &elf_addr, NULL, NULL, NULL);
+		if (!symbol_match(arg, elf_addr, &elf_sym, name))
```

Comment:
`dwfl_module_getsym_info` returns `NULL` on error, so we should check this here for all types of searches rather than in `symbol_match`:

```suggestion
		if (!name || !symbol_match(arg, elf_addr, &elf_sym, name))
```